### PR TITLE
[CodeGen][TII] Allow reassociation on custom operand indices

### PR DESCRIPTION
**Pull request by mshockwave**
This opens up a door for reusing reassociation optimizations on target-specific binary operations with non-standard operand list.

This is effectively a NFC.

### CODE REVIEW

File: llvm/lib/CodeGen/TargetInstrInfo.cpp [hunk 1]
```diff
@@ -1051,43 +1051,45 @@ static std::pair<bool, bool> mustSwapOperands(MachineCombinerPattern Pattern) {
...
-  case MachineCombinerPattern::REASSOC_XA_BY: Row = 2; break;
-  case MachineCombinerPattern::REASSOC_XA_YB: Row = 3; break;
-  default: llvm_unreachable("unexpected MachineCombinerPattern");
-  }
-
-  MachineOperand &OpA = Prev.getOperand(OpIdx[Row][0]);
-  MachineOperand &OpB = Root.getOperand(OpIdx[Row][1]);
-  MachineOperand &OpX = Prev.getOperand(OpIdx[Row][2]);
-  MachineOperand &OpY = Root.getOperand(OpIdx[Row][3]);
+  MachineOperand &OpA = Prev.getOperand(OperandIndices[1]);
```
Comment (topperc) [inline llvm/lib/CodeGen/TargetInstrInfo.cpp:1089]: What does indice 0 represent, I don't see it used but maybe I missed it.

Comment (mshockwave) [reply llvm/lib/CodeGen/TargetInstrInfo.cpp:1089]: It's actually used in `TargetInstrInfo::genAlternativeCodeSequence`, which is `reassociateOps`'s caller, to get the Prev instruction from Root.

File: llvm/lib/CodeGen/TargetInstrInfo.cpp [hunk 2]
```diff
@@ -1184,25 +1248,16 @@ void TargetInstrInfo::genAlternativeCodeSequence(
...
-    Prev = MRI.getUniqueVRegDef(Root.getOperand(1).getReg());
-    break;
-  case MachineCombinerPattern::REASSOC_AX_YB:
-  case MachineCombinerPattern::REASSOC_XA_YB:
-    Prev = MRI.getUniqueVRegDef(Root.getOperand(2).getReg());
-    break;
-  default:
-    llvm_unreachable("Unknown pattern for machine combiner");
-  }
+  std::array<unsigned, 5> OpIdx;
```
Comment (topperc) [inline llvm/lib/CodeGen/TargetInstrInfo.cpp:1251]: Rename to OperandIndices to match what reassociateOps calls it?

File: llvm/lib/CodeGen/TargetInstrInfo.cpp [hunk 3]
```diff
@@ -1184,25 +1248,16 @@ void TargetInstrInfo::genAlternativeCodeSequence(
...
+  getReassociateOperandIdx(Root, Pattern, OpIdx);
```
Comment (topperc) [inline llvm/lib/CodeGen/TargetInstrInfo.cpp:1252]: getReassociateOperandIdx -> getReassociateOperandIndices?

Comment (mshockwave) [reply llvm/lib/CodeGen/TargetInstrInfo.cpp]: Fixed.

File: llvm/lib/CodeGen/TargetInstrInfo.cpp (hunk 2 above)
Comment (mshockwave) [reply llvm/lib/CodeGen/TargetInstrInfo.cpp]: Done

**Comment by mshockwave** [comment]
ping.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**